### PR TITLE
radio contact solution

### PR DESCRIPTION
**Pull request by nwatx**
official solution top down. this one is bottom up.

### CODE REVIEW

File: solutions/USACO jan16-radio-contact.mdx [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+const int INF = 1e9+7;
+const int MX = 1e3+1;
+
+int dd(int a, int b) {
+	return abs(a-b)*abs(a-b);
+}
+
+int dist(pi a, pi b) { // distance squared between two points
+	return dd(a.f, b.f) + dd(a.s, b.s);
+}
```
Comment (bqi343) [inline solutions/USACO jan16-radio-contact.mdx:67]: why not

```cpp
int sq(int x) { return x*x; }
int dist(pi a, pi b) { return sq(a.f-b.f)+sq(a.s-b.s); }
```

File: solutions/USACO jan16-radio-contact.mdx [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+// j = current step of Bessie
+
+int dp[MX][MX];
+
+int main() {
+	setIO("radio");
+
+	cin >> N >> M;
+	
+	for(int i = 0; i < MX; i++) for(int j = 0; j < MX; j++) dp[i][j] = INF;
```
Comment (bqi343) [inline solutions/USACO jan16-radio-contact.mdx:84]: I think `fill_n(dp[0],MX*MX,INF)` works (though I haven't actually done this myself)

File: solutions/USACO jan16-radio-contact.mdx [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+
+	int a, b; cin >> a >> b;
+	vector<pi> jl(N+1); // john location
```
Comment (bqi343) [inline solutions/USACO jan16-radio-contact.mdx:87]: not sure why lines 87,88 are between lines 86,90

**Review by bqi343 — CHANGES_REQUESTED** [review]

File: solutions/USACO jan16-radio-contact.mdx (hunk 2 above)
Comment (nwatx) [reply solutions/USACO jan16-radio-contact.mdx]: it does :)

File: solutions/USACO jan16-radio-contact.mdx [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+// j = current step of Bessie
+
+int dp[MX][MX];
+
+int main() {
+	setIO("radio");
+
+	cin >> N >> M;
+	
+	fill_n(dp[0], MX*MX, INF);
```
Comment (bqi343) [inline solutions/USACO jan16-radio-contact.mdx:84]: I think line 84 should not go in between the reading of `N` and `M` and the initialization of the vectors.

**Review by bqi343 — CHANGES_REQUESTED** [review]

**Review by bqi343 — APPROVED** [review]
code looks fine

File: solutions/USACO jan16-radio-contact.mdx [hunk 5]
```diff
@@ -0,0 +1,129 @@
+---
+id: usaco-598
+source: USACO Gold 2016 January
+title: Radio Contact
+author: Neo Wang
+---
+
+[Official Analysis](http://www.usaco.org/current/data/sol_radio_gold_jan16.html)
+
+## Solution
```
Comment (dongliu0426) [inline solutions/USACO jan16-radio-contact.mdx:10]:
```suggestion
## Explanation
```

File: solutions/USACO jan16-radio-contact.mdx [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+
+Either one of the following happens:
+
+1. Farmer John takes a step $(i+1)$: $\texttt{dp}[i+1][j] = \min(\texttt{dp}[i+1][j], \texttt{dp}[i][j] + \text{dist}(\texttt{jl}[i+1], \texttt{bl}[j]))$
+
+2. Bessie takes a step $(j+1)$: $\texttt{dp}[i][j+1] = \min(\texttt{dp}[i][j+1], \texttt{dp}[i][j] + \text{dist}(\texttt{jl}[i], \texttt{bl}[j+1]))$
+
+3. Both Farmer John and Bessie take steps $(i+1)$ and $(j+1)$: $\texttt{dp}[i+1][j+1] = \min(\texttt{dp}[i+1][j+1], \texttt{dp}[i][j] + \text{dist}(\texttt{jl}[i+1], \texttt{bl}[j+1]))$
+
+## Implementation
```
Comment (dongliu0426) [inline solutions/USACO jan16-radio-contact.mdx:26]:
```suggestion
## Implementation

**Time Complexity:** $\mathcal{O}(N^2)$
```

**Review by dongliu0426 — CHANGES_REQUESTED** [review]
based on https://usaco.guide/general/contributing#adding-a-solution

File: solutions/USACO jan16-radio-contact.mdx [hunk 7]
```diff
@@ -25,6 +25,8 @@ Either one of the following happens:
 
 ## Implementation
 
+**Time Complexity:** $\mathcal{O}(N^2)$
```
Comment (bqi343) [inline solutions/USACO jan16-radio-contact.mdx:28]: NM?

**Review by bqi343 — CHANGES_REQUESTED** [review]

File: solutions/USACO jan16-radio-contact.mdx (hunk 7 above)
Comment (nwatx) [reply solutions/USACO jan16-radio-contact.mdx:28]: woops forgot to change that when copy pasting 😭

**Review by bqi343 — APPROVED** [review]

**Review by dongliu0426 — APPROVED** [review]

**Pull request merged**